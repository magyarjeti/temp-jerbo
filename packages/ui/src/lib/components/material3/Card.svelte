<script lang="ts">
	import { concatClasses } from '@packages/toolbox/concatClasses';
	import StateLayer from './StateLayer.svelte';
	import type { Snippet } from 'svelte';

	interface Props {
		variant?: 'filled' | 'outlined' | 'elevated';
		disabled?: boolean;
		href?: string | null;
		onclick?: (() => void) | null;
		children?: Snippet;
		class?: string | null;
	}

	let { variant = 'filled', href, onclick, children, disabled = false, ...props }: Props = $props();
	let isInteractive = $derived(!!(href || onclick));

	const cardTypeClassMap: Record<typeof variant, string> = {
		filled: 'bg-surface-container-highest',
		elevated: 'bg-surface-container-low elevation-1',
		outlined: 'bg-surface border border-outline-variant',
	};
	const interactiveCardTypeClassMap: Record<typeof variant, string> = {
		filled: 'hover:elevation-1',
		elevated: 'transition hover:elevation-2',
		// TODO
		outlined: '',
	};
	const shapeClass = 'shape-medium';
	const classes = $derived(
		concatClasses(
			props.class,
			'relative overflow-hidden',
			shapeClass,
			cardTypeClassMap[variant],
			isInteractive && interactiveCardTypeClassMap[variant],
			disabled && 'opacity-38',
		),
	);
</script>

{#snippet inside()}
	<StateLayer class={concatClasses('bg-on-surface', shapeClass)} />
	{@render children?.()}
{/snippet}

{#if isInteractive}
	{#if href}
		<a {href} {onclick} class={classes}>
			{@render inside()}
		</a>
	{:else}
		<button {onclick} class={classes}>
			{@render inside()}
		</button>
	{/if}
{:else}
	<div class={classes}>
		{@render children?.()}
	</div>
{/if}
