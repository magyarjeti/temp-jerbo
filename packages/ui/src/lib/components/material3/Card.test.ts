import { render } from '@testing-library/svelte';
import { expect, test } from 'vitest';
import Card from './Card.svelte';

test('Elevated Card snapshot', () => {
	const { container } = render(Card, { props: { variant: 'elevated' } });
	expect(container).toMatchSnapshot();
});

test('Outlined Card snapshot', () => {
	const { container } = render(Card, { props: { variant: 'outlined' } });
	expect(container).toMatchSnapshot();
});

test('Filled Card snapshot', () => {
	const { container } = render(Card, { props: { variant: 'filled' } });
	expect(container).toMatchSnapshot();
});
