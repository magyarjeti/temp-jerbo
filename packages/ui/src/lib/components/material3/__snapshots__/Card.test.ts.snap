// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Elevated Card snapshot 1`] = `
<div>
  <div
    class="overflow-hidden shape-medium bg-surface-container-low elevation-1"
  />
  
</div>
`;

exports[`Filled Card snapshot 1`] = `
<div>
  <div
    class="overflow-hidden shape-medium bg-surface-container-highest"
  />
  
</div>
`;

exports[`Outlined Card snapshot 1`] = `
<div>
  <div
    class="overflow-hidden shape-medium bg-surface border border-outline-variant"
  />
  
</div>
`;
